lantiq: add support for FritzBox 7312 (1&1 WLAN-MODEM)

juppin opened the pull request:
Add support for FritzBox 7312 (1&1 WLAN-MODEM)

HW | Type
----- | -----
SoC: | Lantiq Xway ARX188 PSB 50812 EL
RAM: | 64MB DDR1 (Zentel A4S12D40FTP-G5)
Ethernet: | Atheros 8030
Wireless: | Atheros AR9227 b/g/n 2x2
DSL | Lantiq ADSL2+
DECT | Dialog SC14441
Buttons | WiFi, DECT
LEDs | Power/DSL, Fon, DECT, WLAN, Info

LED | GPIO Number | State
------- | ------- | -------
Power | 44 | active low
Internet | 47 | active low
DECT | 38 | active low
WLAN | 37 | active low
Info | 35 | active low

Button | GPIO Number | State
------- | ------- | -------
WLAN | 1 | active low
DECT | 2 | active low

Ethernet | GPIO Number | State
------- | ------- | -------
25 MHz | 3 | -
Reset | 34 | active low
Interrupt | 39 | -
MII MDIO | 42 | -
MII MDC | 43 | -

PCIE | GPIO Number | State
------- | ------- | -------
Reset | 21 | active low

Information from AVM gpl source: 
[FB7312 GPL](https://osp.avm.de/fritzbox/fritzbox-7312/source-files-FRITZ.Box_7312.06.01.tar.gz)
File: arch/mips/mach-infineon/avm_hw_config_hw189.h

This device is almost the same as FB7330, but only one EHT-Port(at8030) and no USB.
With dt from FritzBox 7330, ethernet did not work.

Fon led is labeled as internet in avm gpl sources.

Main difference to FritzBox 7320 dt:
- Atheros 8030 needs reset (gpio 34) for correct initializing (handled by at803x driver).
- req3 is used as clk for phy instead for pci on fb7320
- No red info led

To install image follow this guide:
https://wiki.openwrt.org/toh/avm/fritz.box.wlan.7330#flash_firmware

Patch for driver/net/ethernet/lantiq_etop.c:
- Patched driver assumes a child in gsw/etop node with name "mdio-bus".
- This is needed for reset gpio definition for at8030 phy.
- Fallback to mdiobus_register() is implemented if no child is defined.

 Node example for patched driver:
```
 &gsw {
	phy-mode = "rmii";
	phy-handle = <&phy0>;
	mtd-mac-address = <&ath9k_cal 0xa91>;
	mtd-mac-address-increment = <(-2)>;

	mdio-bus {
		#address-cells = <1>;
		#size-cells = <0>;
		reg = <0>;

		phy0: ethernet-phy@0 {
			reg = <0>;
			reset-gpios = <&gpio 34 GPIO_ACTIVE_LOW>;
		};
	};
 };
```
Tested with 4.14 kernel version.

Signed-off-by: Johann Neuhauser <johann@it-neuhauser.de>

everloop2 commented:
FRITZ7312.dts looks good

did u check: https://github.com/lede-project/source/pull/1354#pullrequestreview-77612650

LAN "eth0 reset" seems to work in bridged mode (eth0 + wifi) at first sight, if eth0 is set to WAN-DHCP there is NO/sometimes a short reset showing obtained IP for a second (pulling cable/restarting interface)

did a build, kernel 4.9: https://github.com/juppin/openwrt/tree/fb7312 > got no reset (WAN eth0 uplink)

i'll test kernel 4.14

was not able to test: https://github.com/xdarklight/openwrt/commits/fb7312_wip_20160114 cause of missing WLAN and no 2nd LAN port

Greets, everloop2 (aka StefanFreifunk)

everloop2 commented:
kernel 4.14 https://github.com/juppin/openwrt/tree/fb7312 test:

setup: non bridged dhcp-server at wifi and WAN-dhcp-client at eth0 (uplink)

eth0 reset only works if uplink cable is unplugged at boot
-> missing initial boot reset?!

i'll test if its same at kernel 4.9

juppin commented:
I found your PR and read all of the comments.
Didn´t had any of the problems mentioned, except that it does not work if at803x driver gets activated.
Thougt is was resolved due to removing req3 and adding reset-gpios within phy node.

After more testing i can say that if ethernet cable is attached while booting, i do not have any problems.
If i attach ethernet cable after boot up, approximately one out of three test results in no working phy until next reboot.
This is complete inverse to your description of no working phy.
I had also tested bridged and dedicated lan and wan port, but this makes no difference on my device.

How can you say, that phy did a hard reset?

I had added some printk´s in at803x.c to test if driver gets the right gpio and did a reset on link change...
Line [262](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/tree/drivers/net/phy/at803x.c?h=linux-4.14.y#n262) always return a null pointer with dt from my pr ...
`gpiod_reset = devm_gpiod_get_optional(dev, "reset", GPIOD_OUT_LOW);`

Thats why line [342](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/tree/drivers/net/phy/at803x.c?h=linux-4.14.y#n342) is always false and no hard reset is triggered.
`if (priv->gpiod_reset && !priv->phy_reset) {`

So reset is not working because that gpiolib did not find the reset-gpios property in device node.
The only example with reset-gpios node for a phy in a dt that i found is from [Fritzbox 7360SL](https://github.com/openwrt/openwrt/blob/ac593d895a00206e744ed8288d434e439b069e03/target/linux/lantiq/files-4.14/arch/mips/boot/dts/FRITZ7360SL.dts#L122).
But the dt tree source vr9.dtsi is very different to ar9.dtsi

Had also tried to add reset-gpios right after mtd-mac-...

Could anyone help with correct device node definition???

mkresin commented on target/linux/lantiq/files-4.14/arch/mips/boot/dts/FRITZ7312.dts:
```diff
@@ -0,0 +1,183 @@
+		poll-interval = <100>;
+
+		rfkill {
+			label = "rfkill";
```
the label on the case says "WLAN". Please stick to this name.

mkresin commented on target/linux/lantiq/base-files/etc/board.d/01_leds:
```diff
@@ -53,6 +53,9 @@ bt,homehub-v5a)
 netgear,dm200)
 	ucidef_set_led_netdev "lan" "lan" "dm200:green:lan" "eth0"
 	;;
+avm,fritz7312)
```
This shouldn't be required. Due to the `led-wifi` alias in the dts, the `phy0tpt` trigger should be used by default.

mkresin commented:
@StefanFreifunk would it be possible that you have a look at this PR. Might be the issue you had is fixed here.

juppin commented:
Requested changes are done

everloop2 commented:
@juppin: How can you say, that phy did a hard reset?
Dont know for sure if phy does a reset or not at boot - i'll watch for it.

https://patchwork.ozlabs.org/patch/361357/

Have seen other boxes, short before boot finished, phy or network is "reset".
(link gets up while booting, on some boxes without assigning IP - think due DHCP still down, short reconnect, IP assigned, everything working)

7312 i play with: get a working phy by unplugging cable at boot i've seen first time yesterday, hope its not faulty hardware

7330SL also has at8030 -> 2nd phy (not working), next task - see if can test it there

https://github.com/xdarklight/openwrt/commit/b63e57fca57f7f00815d21feac6519fd564532d1, FRITZ7312.dts:

			phy0 {
				reg = <0>;
				reset-gpios = <&gpio 34 GPIO_ACTIVE_LOW>;
				interrupt-parent = <&eiu0>;
				interrupts = <39>;                      <- or <40>
			};

i'am no coder and quite new to openwrt


this is where i stuck:

		etop@E180000 {
			#address-cells = <1>;
			#size-cells = <0>;
			phy-mode = "rmii";
			lantiq,mdc-clock-rate = <1000000>; /* 1MHz */
			mtd-mac-address = <&ath9k_cal 0xa91>;
			mtd-mac-address-increment = <(-2)>;

			phy0 {
				reg = <0>;
				reset-gpios = <&gpio 34 GPIO_ACTIVE_LOW>;
				interrupt-parent = <&eiu0>;
				interrupts = <39>;
			};
		};

juppin commented:
> Dont know for sure if phy does a reset or not at boot - i'll watch for it.

i´ve added a patch file that should be placed in target/lantiq/patch-4.X that prints some simple info...
[0001-add-reset-gpio-debug-output.zip](https://github.com/openwrt/openwrt/files/1871242/0001-add-reset-gpio-debug-output.zip)
Look with `dmesg | grep at803x` if reset gpio is resolved from dt (gpiod_desc should not be 0) and if this are working, driver should do a hard reset if phy gets down.

You could test your hw if you flash stock avm firmware... [FRITZ.Box_7312.117.06.54.image](http://ftp.avm.de/fritzbox/fritzbox-7312/deutschland/fritz.os/FRITZ.Box_7312.117.06.54.image)

[https://patchwork.ozlabs.org/patch/361357/](https://patchwork.ozlabs.org/patch/361357/) same as [reset-gpios mainline commit at803x.c](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/drivers/net/phy/at803x.c?id=13a56b4493259e6b020dbcf6a76b2bef479f3edf)

Due to my device doesn´t resolv gpio from dt, i assume phy node isn´t correct or on the wrong place.
Could be that phy node should not placed as child of `gsw: etop@e180000` or with a mido subnode as on FB7360SL. 

As described here ([dt bindings phy](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/net/phy.txt)), reg, interrupt-parent and interrups are required for phy dt node and should be correct.


We can fix this for our own use with modifying the driver to resolve the needed gpio as an system gpio, but this would be very ugly and would never get mainline to openwrt.

everloop2 commented:
mdio i've seen in at803x driver: https://github.com/torvalds/linux/blob/master/drivers/net/phy/at803x.c

found 3 patches fixing some issues: http://lists.infradead.org/pipermail/lede-commits/2016-December/001651.html,

/pending-4.4/736-at803x-fix-reset-handling.patch
/pending-4.4/737-net-phy-at803x-Request-reset-GPIO-only-for-AT8030-PH.patch
/pending-4.4/738-net-phy-at803x-only-the-AT8030-needs-a-hardware-rese.patch


at ../source/kernel/ref-ar9-7320_06.30/linux-2.6.32/drivers/net/avm_cpmac/configs/config_7312.h found:

```
#include <avmnet_module.h>
#include <avmnet_config.h>
#include "../phy/avmnet_ar803x.h"
#include "../switch/ifx/ar9/swi_ar9.h"
#include "../switch/ifx/ar9/mac_ar9.h"

.initdata.mac       = { 
                            .mac_nr = 0,
                            .mac_mode = MAC_MODE_RMII	> RMII
                            .rx_delay = 0x82ee,
                            .tx_delay = 0x2d47,
                            .mdio_addr = 0x00,		> mdio?!
                            .reset = 34,		> GPIO 34 reset
                            .lnkf_time = 2 * HZ,	> ??	2 MHz?
                            .irq = IFX_EIU_IR3,
                          },
```

everloop2 commented:
did a test build with empty "state_default: pinmux" [FRITZ7312.dts.zip](https://github.com/openwrt/openwrt/files/1871546/FRITZ7312.dts.zip),
same bug - if lan cable at boot is unplugged, reset is working

juppin commented:
Hello back,

sorry for this delay, i was very busy last two weeks.

I got reset of phy through at803x driver working with help of @mkresin .
Kernel 4.9 was no so difficult, kernel 4.14 needed some extra options in dt.

But now, i think device trees of other devices that uses etop driver need some modification to work with this patch for etop driver.

What is the preferred solution? 

1.  Fallback to mdiobus_register if of_mdiobus_register get not called because nothing is defined in dt

1. Update dt of other devices to have a mdio node with phy(s)

And if this patch for etop driver get accepted, should it be a extra patch file or integrated in the various etop fixes patch?

PR istn´t ready for merging!!! 

@everloop2 Could you verify if that pr does work on your device too? You could include the patch file for simple debug output if you want to see output of successful reset when cable gets unplugged...

juppin commented:
Rebased on latest head and tested for over four days as dump-ap.
I had tested more different solutions to get reset working, but this seems the easiest and cleanest one.
If additional information is required feel free to ask a question...

juppin commented:
My device only reset the phy if link go´s down...

If i disconnect lan cable i get this message:
`[  665.207397] at803x: phy has no link -> reset (gpiod_reset=0x838a4220, phy_reset=0x0)`
And while cable is disconnected i get every second this message until i connect the lan cable:
`[  683.711283] at803x: phy has no link -> no reset (gpiod_reset=0x838a4220, phy_reset=0x1)`

Probably your other eth device misbehaves? Could you test it with a other device?

juppin commented:
Sorry i can not reproduce your mentioned behavior...

Why are you always mention WAN?
Had you configured the lan port as wan or did you mean dsl modem?

I had tested this eth port with at8030 phy only as configured lan.

My different test setups are:

- FB7312 as Dump AP with lan port (connected to other router) configured as dhcp client and bridged with wlan
- FB7312 as wireless client  (connected to other router) configured as wwan and eth port configured as lan with dhcp (connected to notebook)

If i disconnect lan cable or does restart network with init script, phy reset happens only once a time.

Could you describe your test setup to reproduce your mentioned behavior?
Could you also test FB7312 eth port configured as lan with dhcp and connected to pc or notebook?

jneuhauser commented:
@everloop2 I need your feedback or your device ;-)

I have updated my fb7312 branch:
1. Rebased on latest changes. 

1. Fallback to midobus_register implemented if no subnode with needed compatible property is defined in dt.

So no device should be broken with this changes!

How is the chance that FB7312 got merged?
Do we need more testers?

everloop2 commented:
tried -> https://github.com/juppin/openwrt/tree/fb7312

unfortunately wifi does not work anymore both kernel 4.9 and 4.14, not even shown in logs -> no owl-loader (wifi firmware) -> see: https://openwrt.org/toh/avm/fritz.box.wlan.7312#bootlogs

my test setup:
FB7312 as lan dhcp-client (eth0 "lan-port" connected to cable modem) and wifi AP (dhcp server)

when first build stock openwrt applying minor changes (https://openwrt.org/toh/avm/fritz.box.wlan.7312#firmware_build_fixes) everything seemed to work -> lan&wifi in **bridged** mode -> at8030 does reset,

for my usecase needed eth0 (lan-port) as WAN (dhcp-client), lan&wifi **not bridged** -> at8030 does not reset anymore see: https://github.com/openwrt/openwrt/pull/840#issuecomment-378023746

thanks for ur work :)

juppin commented:
> unfortunately wifi does not work anymore both kernel 4.9 and 4.14, not even shown in logs -> no owl-loader (wifi firmware) -> see: https://openwrt.org/toh/avm/fritz.box.wlan.7312#bootlogs

dont´t know whats wrong with your build... for me wifi and eth does work as expected with both kernel versions... i also see owl-loader in kernel log...

i had rebased on latest HEAD and dropped also kernel 4.9 support

jneuhauser committed:
Add support for FritzBox 7312 (1&1 WLAN-MODEM)

HW	| Type
------- | ------------------------------------
SoC	| Lantiq Xway ARX188 PSB 50812 EL
RAM	| 64MB DDR1 (Zentel A4S12D40FTP-G5)
Ethernet| Atheros 8030 (2x PEF7071V on FB7320)
Wireless| Atheros AR9227 b/g/n 2x2
DSL	| Lantiq ADSL2+
DECT	| Dialog SC14441
Buttons	| WiFi, DECT
LEDs	| Power/DSL, Fon, DECT, WLAN, Info

LED	| GPIO	| State
------- | ----- | -----------
Power	| 44	| active low
Internet| 47	| active low
DECT	| 38	| active low
WLAN	| 37	| active low
Info	| 35	| active low

Button	| GPIO	| State
------- | ----- | -----------
WLAN	| 1	| active low
DECT	| 2	| active low

Phy	| GPIO	| State
------- | ----- | -----------
25 MHz	| 3	| -
Reset	| 34	| active low
Int	| 39	| -
MII MDIO| 42	| -
MII MDC	| 43	| -

PCIE	| GPIO	| State
------- | ----- | -----------
Reset	| 21	| active low

Information from AVM gpl source:
Url:	https://osp.avm.de/fritzbox/fritzbox-7312/source-files-FRITZ.Box_7312.06.01.tar.gz
File:	arch/mips/mach-infineon/avm_hw_config_hw189.h

This device is almost the same as FB7330, but only one EHT-Port and no USB.
With dt from FritzBox 7330, ethernet did not work.

Fon led is labeled as internet in avm gpl sources.

Main difference to FritzBox 7320 dt:
 - Atheros 8030 needs reset (gpio 34) for correct initializing (handled by at803x driver).
 - req3 is used as clk for phy instead for pci on fb7320
 - No red info led

Patch for driver/net/ethernet/lantiq_etop.c:
 This driver assumes a child in gsw/etop node with the name "mdio-bus".
 This is needed for reset gpio definition for at8030 phy.
 Fallback to mdiobus_register() is implemented if no child is defined.

 Example:
 &gsw {
	phy-mode = "rmii";
	phy-handle = <&phy0>;
	mtd-mac-address = <&ath9k_cal 0xa91>;
	mtd-mac-address-increment = <(-2)>;

	mdio-bus {
		#address-cells = <1>;
		#size-cells = <0>;
		reg = <0>;

		phy0: ethernet-phy@0 {
			reg = <0>;
			reset-gpios = <&gpio 34 GPIO_ACTIVE_LOW>;
		};
	};
 };

To install image follow this guide:
https://wiki.openwrt.org/toh/avm/fritz.box.wlan.7330#flash_firmware

Signed-off-by: Johann Neuhauser <johann@it-neuhauser.de>

everloop2 commented:
tried ur image:
-> FB7312 as lan dhcp-client (eth0 "lan-port" connected to router) and wifi AP (dhcp server)
-> unplugged/plugged lan min. 4 times -> "ifconfig" showed eth0 ip after a few secs, got connection & internet -> seems to be workin :D

i'll try a fresh build -> https://github.com/juppin/openwrt.git -> fb7312 branch + LuCI

everloop2 commented:
fb7312 branch builds flawless - got a workin image :D

nevertheless something is borked with my box, flashed a workin image including debug patch a few times -> "dmesg | grep at803x" showed reset but no ip is obtained -> ill try another device (FritzBox7330SL - 2nd LAN is at8030)

juppin commented:
> nevertheless something is borked with my box

I think also there is something broken with your wan modem...
I also have a wan cable moden, but for me it is no usecase to drive this device as main router due to only one ethernet port.
This device is great for a dump ap with probably some services as voip to analog bridge...

juppin commented:
@mkresin Any suggestions for improvements?
If you are okay with the etop patch, i will include it in the 0028-NET-lantiq-various-etop-fixes.patch

Would like to see this device on 18.06 :-)

Klapstuhl commented:
@juppin As you said this device really is great as a AP / VoIP bridge. Is it possible to use the ltq-tapi / ltq-vmmc modules with this device? I want to use it with Asterisk15-chan-lantiq but I haven't figured out how...

juppin commented:
@Klapstuhl Yes is possible...
I´ve used asterisk13 with that packages:
```
asterisk13-app-stack
asterisk13-chan-lantiq
asterisk13-chan-sip
asterisk13-codec-a-mu
asterisk13-codec-alaw
asterisk13-codec-ulaw
asterisk13-res-rtp-asterisk
```

Here are some config exaples... https://forum.lede-project.org/t/voip-configuration-for-asterisk13-pjsip-chan-lantiq-and-vodafone-germany/9470

Klapstuhl commented:
I just tried to build OpenWRT with ltq-tapi, ltq-vmmc and the asterisk13 stuff you mentioned. 
The build completes with problems but the FritzBox keeps restarting after the Lantiq driver is loaded.
Idk how to solve this problem tbh.

mkresin commented:
Hey @juppin,

I've to admit, that I have completely forgotten your PR. I merged your PR to [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary) and applied the following changes:

- etop code changes moved to extra commit
- removed usb modules from the package selection
- fixed some spaces vs. tabs
- prefix added to commit title
- rewrote commit message to get rid of the github markup
- dropped some unrelated informations from the commit message
- added installation guide to the commit message

Do you agree with the changes I've done? Does the applied changes make sense to you? Maybe it would be a good idea to runtime test an image build from my staging tree.

juppin commented:
@mkresin Your changes look very good and makes sense to me. Will build and test a image from your staiging tree today in the afternoon.
Thanks

juppin commented:
@mkresin Works as expected. Could be merged in master.

There is a small error in commit message of etop driver patch, you write "Fallback to mdiobus_register() if mdio-bus child node exists." but it should be "Fallback to mdiobus_register() if mdio-bus child node doesn´t exists."

Thanks

mkresin commented:
Commit message fixed and pushed to master. Thanks a lot for your work and patience.
